### PR TITLE
Added maxWidth and maxHeight property to the cropper

### PR DESCRIPTION
**Pull request by AmoghHegde**
Implemented cropperMaxWidth and cropperMaxWidth property which can provided as an input in the API and this ensures that the cropper width and height cannot be increased more than the value provided in the input.

If no value is provided, it is set to image width and height.

### CODE REVIEW

**Comment by Spymannn** [comment]
anyone knows when this branch will be merge ? 

Thanks

**Comment by itreppert** [comment]
Yeah, would be very nice to have it. @Mawi137
Really need it in a customers project.

**Comment by Spymannn** [comment]
Still waiting for this branch to be merged... @Mawi137 
any news about it ? is this project "dead" ?

**Comment by GuilhermeDemoliner** [comment]
Still waiting for this branch to be merged to.... @Mawi137 
Thanks.

**Comment by Spymannn** [comment]
@Mawi137 ? any news ? is this repo dead or what ?

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts [hunk 1]
```diff
@@ -415,23 +425,25 @@ export class ImageCropperComponent implements OnChanges, OnInit {
             this.cropper.y2 = sourceImageElement.offsetHeight > this.cropperStaticHeight ?
                 this.cropperStaticHeight : sourceImageElement.offsetHeight;
         } else {
+            const cropperWidth = Math.min(this.cropperScaledMaxWidth,sourceImageElement.offsetWidth);
+            const cropperHeight = Math.min(this.cropperScaledMaxHeight, sourceImageElement.offsetHeight);
             if (!this.maintainAspectRatio) {
```
Comment (Mawi137) [inline projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:430]: Normally the cropper takes up the full image and if there's an aspect ratio, it will center the cropper. Would be nice to keep it that way when there's a max height or width set.

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts [hunk 2]
```diff
@@ -615,6 +627,25 @@ export class ImageCropperComponent implements OnChanges, OnInit {
         }
     }
 
+    private setCropperScaledMaxSize(): void {
+        const sourceImageElement = this.sourceImage.nativeElement;
+        const ratio = this.transformedSize.width / sourceImageElement.offsetWidth;
```
Comment (Mawi137) [inline projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:632]: `this.maxSize.width` can be used instead of `sourceImageElement.offsetWidth`, they should be the same.

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts [hunk 3]
```diff
@@ -615,6 +627,25 @@ export class ImageCropperComponent implements OnChanges, OnInit {
...
+        this.cropperScaledMaxWidth = this.cropperMaxWidth > 20? this.cropperMaxWidth : this.maxSize.width;
```
Comment (Mawi137) [inline projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:633]: This part is quite hard to follow, could you make it clearer in the code what should happen if there is a max height or width set? It also seems you assume both will be set, which is not always the case (ratio is applied to both even if only one is set).

Or, why not follow the same logic as for `cropperMinWidth` and `cropperMinHeight`?

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts [hunk 4]
```diff
@@ -661,47 +692,63 @@ export class ImageCropperComponent implements OnChanges, OnInit {
     }
 
     private resize(event: any): void {
-        const diffX = this.getClientX(event) - this.moveStart.clientX;
-        const diffY = this.getClientY(event) - this.moveStart.clientY;
+        var diffX = this.getClientX(event) - this.moveStart.clientX;
```
Comment (Mawi137) [inline projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:695]: var in TypeScript?

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts [hunk 5]
```diff
@@ -661,47 +692,63 @@ export class ImageCropperComponent implements OnChanges, OnInit {
...
                 const scale = event.scale;
                 const newWidth = (Math.abs(this.moveStart.x2 - this.moveStart.x1)) * scale;
                 const newHeight = (Math.abs(this.moveStart.y2 - this.moveStart.y1)) * scale;
                 const x1 = this.cropper.x1;
                 const y1 = this.cropper.y1;
-                this.cropper.x1 = Math.min(this.moveStart.clientX - (newWidth / 2), this.cropper.x2 - this.cropperScaledMinWidth);
-                this.cropper.y1 = Math.min(this.moveStart.clientY - (newHeight / 2), this.cropper.y2 - this.cropperScaledMinHeight);
-                this.cropper.x2 = Math.max(this.moveStart.clientX + (newWidth / 2), x1 + this.cropperScaledMinWidth);
-                this.cropper.y2 = Math.max(this.moveStart.clientY + (newHeight / 2), y1 + this.cropperScaledMinHeight);
+                this.cropper.x1 = Math.min(Math.max(this.moveStart.clientX - (newWidth / 2), this.cropper.x2 - this.cropperScaledMaxWidth/2),
```
Comment (Mawi137) [inline projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:744]: Shouldn't `this.moveStart.clientX` be used instead of `this.cropper.x2`? `clientX` is the middle of the image before pinch zooming (on mobile).

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts [hunk 6]
```diff
@@ -798,8 +845,15 @@ export class ImageCropperComponent implements OnChanges, OnInit {
         if (this.sourceImage && this.sourceImage.nativeElement && this.transformedImage != null) {
             this.startCropImage.emit();
             const imagePosition = this.getImagePosition();
-            const width = imagePosition.x2 - imagePosition.x1;
-            const height = imagePosition.y2 - imagePosition.y1;
+
+            const imgWidth = imagePosition.x2 - imagePosition.x1;
```
Comment (Mawi137) [inline projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:849]: Is all this new code really necessary? I don't think `imgWidth` can ever be bigger than `scaledMaxWidthWithRatio`.

**Review by Mawi137 — CHANGES_REQUESTED** [review]

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts (hunk 3 above)
Comment (AmoghHegde) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:633]: When maxWidth or maxHeight is not provided in the input, cropperScaledMaxWidth/cropperScaledMaxHeight will be set to maxSize width/height respectively and adjusted according to the aspect ratio. 
For setting the cropperMinWidth/cropperMinHeight, the cropperMinWidth which is provided in input is assigned and the cropperMinHeight is set based on aspect ratio. 
Here, we check whether the cropperScaledMaxWidth/cropperScaledMaxHeight should be reduced based on aspect ratio since neither of them should be more than what is provided in the input.

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts (hunk 4 above)
Comment (AmoghHegde) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:695]: Change to const

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts (hunk 6 above)
Comment (AmoghHegde) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:849]: Not needed. Remove the new code.

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts (hunk 2 above)
Comment (AmoghHegde) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:632]: Use this.maxSize.width instead of sourceImageElement.offsetWidth

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts (hunk 5 above)
Comment (AmoghHegde) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:744]: Change this.cropper.x2 to this.moveStart.clientX

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts (hunk 1 above)
Comment (AmoghHegde) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:430]: How do you recommend I do that?
I have not tested this feature extensively on mobile devices.

Comment (AmoghHegde) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:435]: How do you recommend I do that?

Comment (AmoghHegde) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:435]: Center the cropper when there is max height or width set

Comment (AmoghHegde) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:435]: Done

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts (hunk 4 above)
Comment (AmoghHegde) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts]: Done

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts (hunk 6 above)
Comment (AmoghHegde) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts]: Done

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts (hunk 5 above)
Comment (AmoghHegde) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts]: Done

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts (hunk 2 above)
Comment (AmoghHegde) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts]: Done

File: projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts (hunk 1 above)
Comment (AmoghHegde) [reply projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:435]: Done

**Comment by AmoghHegde** [comment]
I have fixed the code review comments and resolved the merge conflict. Thanks.

**Review by Mawi137 — APPROVED** [review]

**Comment by Mawi137** [comment]
Thanks! Released in version `3.3.0`.
